### PR TITLE
Feature newapp shallow water

### PR DESCRIPTION
**Pull request by miguelmaso**
this is the application I have been working on
It has four elements to solve the shallow water equations: primitive or conserved variables, and pfem2 or eulerian framework.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -56,14 +56,4 @@ kratos/includes/kratos_version.h
...
-*.h~
-
-applications/FluidDynamicsApplication/test_examples/fs_cavity/ProjectParameters\.py
-
-applications/FluidDynamicsApplication/test_examples/fs_cavity/run_test\.py
-
-*.cpp~
-
-applications/python_scripts/beam_profiles\.csv
+cmake_debug/
```
Comment (jcotela) [inline .gitignore:59]: Please don't add non-standard paths and files to `.gitignore`. You can get git to ignore specific files or folders in your machine by modifying the file `.git/info/exclude`, found in the hidden `.git` directory in the kratos folder.

See [this](https://github.com/KratosMultiphysics/Kratos/wiki/How-to-make-Git-ignore-the-files-resulting-from-a-compilation-without-conflicts-in-.gitignore) wiki entry for details.

File: applications/CMakeLists.txt [hunk 2]
```diff
@@ -192,9 +193,13 @@ if(${PFEM2_APPLICATION} MATCHES ON)
...
-if(${PURE_DIFFUSION_APPLICATION} MATCHES ON)
-  add_subdirectory(pure_diffusion_application)
-endif(${PURE_DIFFUSION_APPLICATION} MATCHES ON)
+if(${MACHINING_APPLICATION} MATCHES ON)
+  add_subdirectory(MachiningApplication)
+endif(${MACHINING_APPLICATION} MATCHES ON)
+
+if(${STRING_DYNAMICS_APPLICATION} MATCHES ON)
+  add_subdirectory(StringDynamicsApplication)
+endif(${STRING_DYNAMICS_APPLICATION} MATCHES ON)
```
Comment (jcotela) [inline applications/CMakeLists.txt:202]: If I understand this correctly, you are modifying the add_subdirectory commands for several applications, but you are only adding the `SHALLOW_WATER_APPLICATION`. Could you revert the changes to other applications? Thanks!

File: applications/ShallowWaterApplication/CMakeLists.txt~ [hunk 3]
```diff
@@ -0,0 +1,46 @@
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
```
Comment (jcotela) [inline applications/ShallowWaterApplication/CMakeLists.txt~:1]: This file is a backup (note the trailing `~`) and should not be added to the repository

File: applications/ShallowWaterApplication/TestApplication.py~ [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from KratosTestApplication import *
```
Comment (jcotela) [inline applications/ShallowWaterApplication/TestApplication.py~:1]: Also a backup, please remove.

File: applications/ShallowWaterApplication/custom_conditions/nothing_condition.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+==============================================================================
+ */
+
+//
+//   Project Name:        Kratos
+//   Last modified by:    Miguel Masó Sotomayor
+//   Date:                September 27th 2017
+//   Revision:            1.0
+//
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_conditions/nothing_condition.cpp:47]: This is the old header, going forward, I'd recommend you to use the one in [this template](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/templates/header_template):
```
//    |  /           |
//    ' /   __| _` | __|  _ \   __|
//    . \  |   (   | |   (   |\__ `
//   _|\_\_|  \__,_|\__|\___/ ____/
//                   Multi-Physics
//
//  License:                BSD License
//                               Kratos default license: kratos/license.txt
//
//  Main author:    Author1 Fullname
//
```

File: applications/ShallowWaterApplication/custom_conditions/nothing_condition.cpp [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+CLAIM, DAMAGES OR  OTHER LIABILITY, WHETHER IN AN  ACTION OF CONTRACT,
+TORT  OR OTHERWISE, ARISING  FROM, OUT  OF OR  IN CONNECTION  WITH THE
+SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+==============================================================================
+ */
+
+//
+//   Project Name:        Kratos
+//   Last modified by:    Miguel Masó Sotomayor
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_conditions/nothing_condition.cpp:44]: Unfortunately, tildes, accents and other non-ascii characters tend to cause problems when changing between windows and linux (they become corrupted, depending on the editor used). I'd ask you to use a non-accented version of your name in h files.

File: applications/ShallowWaterApplication/custom_elements/conserved_var_element.cpp [hunk 7]
```diff
@@ -0,0 +1,384 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_elements/conserved_var_element.cpp:1]: See comments on header.

File: applications/ShallowWaterApplication/custom_elements/conserved_var_element.cpp [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+    {
+        KRATOS_TRY
+
+        const GeometryType& rGeom = this->GetGeometry();
+        const PropertiesType& rProp = this->GetProperties();
+
+        // verify nodal variables and dofs
+        for ( unsigned int i = 0; i < TNumNodes; i++ )
+        {
+            // Verify basic variables
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_elements/conserved_var_element.cpp:79]: A very minor comment, for your convenience: you may want to use the macros
```
KRATOS_CHECK_VARIABLE_KEY(TheVariable)
KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(TheVariable, TheNode)
KRATOS_CHECK_DOF_IN_NODE(TheVariable, TheNode)
```
in [checks.h](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/checks.h), which have been designed to do this kind of checks.

File: applications/ShallowWaterApplication/custom_elements/euler_cons_var_element.cpp [hunk 9]
```diff
@@ -0,0 +1,439 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_elements/euler_cons_var_element.cpp:1]: See notes on header.

File: applications/ShallowWaterApplication/custom_elements/euler_cons_var_element.cpp [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+        {
+            rdepth[counter] = 0;
+            runkn[counter]  = rGeom[i].FastGetSolutionStepValue(MOMENTUM_X);
+            rprev[counter]  = rGeom[i].FastGetSolutionStepValue(MOMENTUM_X,1);
+            counter++;
+
+            rdepth[counter] = 0;
+            runkn[counter]  = rGeom[i].FastGetSolutionStepValue(MOMENTUM_Y);
+            rprev[counter]  = rGeom[i].FastGetSolutionStepValue(MOMENTUM_Y,1);
+            counter++;
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_elements/euler_cons_var_element.cpp:355]: You could optimize this by getting the MOMENTUM variable once (as a vector) and accessing its components.

File: applications/ShallowWaterApplication/custom_python/add_custom_strategies_to_python.h [hunk 11]
```diff
@@ -0,0 +1,85 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_python/add_custom_strategies_to_python.h:1]: See comments on header

File: applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 12]
```diff
@@ -0,0 +1,108 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.cpp:1]: Header again...

File: applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.h [hunk 13]
```diff
@@ -0,0 +1,78 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_python/add_custom_utilities_to_python.h:1]: Header again...

File: applications/ShallowWaterApplication/custom_python/shallow_water_python_application.cpp [hunk 14]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_python/shallow_water_python_application.cpp:1]: Header again...

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 15]
```diff
@@ -0,0 +1,2288 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h:1]: Header...

File: applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h [hunk 16]
```diff
@@ -0,0 +1,2288 @@
...
+            {
+                ModelPart::NodesContainerType::iterator inode = inodebegin+ii;
+                noalias(inode->GetSolutionStepValue(OriginVariable,1)) = inode->FastGetSolutionStepValue(OriginVariable);
+            }
+        }
+        KRATOS_CATCH("")
+    }
+
+
+    //~ void ComputeVelocity(ModelPart::NodesContainerType& rNodes)
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_utilities/move_shallow_water_particle_utility.h:513]: I'd recommend removing commented out functions in the future (the idea is that you can always access the git history to see old versions of the code)

File: applications/ShallowWaterApplication/custom_utilities/shallow_water_particle.h [hunk 17]
```diff
@@ -0,0 +1,180 @@
+
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_utilities/shallow_water_particle.h:1]: header...

File: applications/ShallowWaterApplication/custom_utilities/shallow_water_particle.h [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+    ///@{
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    ShallowParticle(TDataType const& NewX, TDataType const& NewY, TDataType const& NewZ) : Point(NewX, NewY, NewZ)
+    {
+        this->ERASE_FLAG = true; //initializing as useless particle
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_utilities/shallow_water_particle.h:134]: try to avoid all caps names for variables (we reserve them for Kratos::Variable objects)

File: applications/ShallowWaterApplication/custom_utilities/shallow_water_variables_utility.h [hunk 19]
```diff
@@ -0,0 +1,264 @@
+/*
```
Comment (jcotela) [inline applications/ShallowWaterApplication/custom_utilities/shallow_water_variables_utility.h:1]: header...

File: applications/ShapeOptimizationApplication/TestApplication.py~ [hunk 20]
```diff
@@ -0,0 +1,10 @@
+from KratosTestApplication import *
```
Comment (jcotela) [inline applications/ShapeOptimizationApplication/TestApplication.py~:1]: I think this file should not be here (it is a backup)

File: applications/applications_interface.py [hunk 21]
```diff
@@ -44,6 +44,8 @@
 Import_MultiScaleApplication = False
```
Comment (jcotela) [inline applications/applications_interface.py:44]: applications_interface is no longer used, we have it here for backwards compatibility (and I think we should remove it soon), but you don't need to modify it.

**Review by jcotela — CHANGES_REQUESTED** [review]
I've added several comments, most are cosmetic. The only thing I'd ask you to approve this is to remove the changes on gitignore, the application lists and backup files.

The code itself looks good to me (keeping into account that I haven't tested it). For the future, I'd recommend taking a look into the kratos style conventions for variable names (https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide)

**Comment by miguelmaso** [comment]
I have found a bug in my last commit. I am going to:
- Close this PR
- Fix the bug
- Make some of the requested changes
- Open another PR
